### PR TITLE
2250/auth cookie

### PR DESCRIPTION
**Pull request by dominikx96**
# Pull Request Template

## Issue Overview

This PR addresses #2250

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Refactors authentication flow from session storage to a secured cookie. The reason why we should use the cookie, instead of other options is described here: https://tkacz.pro/how-to-securely-store-jwt-tokens/

I've refactored backend & frontend apps. Once a user tries to authenticate, a backend creates a cookie with httpOnly and secure options and sends along with a response. The cookie is visible in browser cookies tab, but there is no way to read the cookie using javascript.

From the frontend perspective, to authorize requests I had to add the `withCredentials` parameter to an axios config. It sends cookies along with a request.

To log out a user, I defined a new `/logout` endpoint that creates a new cookie with the current date and sends in response. Because a cookie date is set to the current date, it means it's expired to a user is logged out.

## How Can This Be Tested/Reviewed?
The easiest way to test this is to go through a few of the different log in processes

try to log into the public site (as a non-mfa user) then try opening a new tab. You should remain logged in to both without needing to relog in

try to log into the partner site (as an mfa user and later as a nonmfa user). Try opening a new tab, and you should see that user remains logged in the new tab. 

you should also try to make sure that if the user is a jurisdictional admin then the permissions should still work correctly 

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by seanmalbert** [comment]
@dominikx96 and @YazeedLoonat,

I made an update to this PR to use an allowlist defined as an environment variable. See `backend/core/.env.template`'s `CORS_ORIGINS` for the value to set in your `.env`

File: backend/core/src/auth/dto/login-response.dto.ts
Comment (seanmalbert) [inline backend/core/src/auth/dto/login-response.dto.ts:6]: Do we need this or can we use StatusDto in its place?

File: backend/core/src/auth/dto/logout-response.dto.ts
Comment (seanmalbert) [inline backend/core/src/auth/dto/logout-response.dto.ts:6]: same question here, can we use StatusDto instead?

**Comment by seanmalbert** [comment]
> Another problem is with refreshing a token. I'm not sure if I understand correctly, but it looks like the /auth/token is responsible for refreshing a token, am I right?

@dominikx96,
Yes, /auth/token calls `generateAccessToken`, so I think we can add the refresh token logic to something that gets hit on every request, like AuthzGuard's canActivate. Would like support from @YazeedLoonat to implement that piece here?

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
Looks good, but two small issues I've found:
when you first hit the public and when you first hit the partner site we get an `UnauthorizedException` which I think is going to bog down papertrail on production/dev sites

**Comment by dominikx96** [comment]
@seanmalbert @YazeedLoonat 

About `UnauthorizedException` - we need to know if a user is logged in to show proper navigation elements so that's why I'm calling the user data endpoint. When we had a token in local storage, we could check if a token exists, but currently, there is no way to read a secured cookie. I don't have any idea how we can prevent `UnauthorizedException` and check if a user is logged in.

I'm also not sure about refresh token logic - a refresh token should be longer than an access token, but don't know what exp date we should define for both. @YazeedLoonat can you help me with this part? I can assist you and implement the frontend part! :) 

My vision here is: once a user gets a message that the token is expired, we can call the refresh token endpoint, if a refresh token is valid, then we can update a cookie and a user is logged in, if not - we can update a cookie with exp time set to actual time, so it means - a user will be logged out.

**Comment by YazeedLoonat** [comment]
So updates on here should be good to go! I got the backend there and got the front end to call the new refreshToken as well 

So I think this should all be squared away

**Comment by ludtkemorgan** [comment]
@YazeedLoonat this PR doesn't start up because of an error in app.module.ts. Is that a known issue?

**Comment by YazeedLoonat** [comment]
update: there is a .env change in backend/core so for reviewers please do include those changes!

File: backend/core/.env.template
Comment (ludtkemorgan) [inline backend/core/.env.template:21]: just curious what the 9000 port is for?

File: backend/core/src/app.module.ts
Comment (ludtkemorgan) [inline backend/core/src/app.module.ts:44]: might be a good idea just in case we miss a config file configuration in one of the environments we should change this line to 
`const allowList = process.env.CORES_ORIGINS || []`

File: backend/core/src/auth/controllers/auth.controller.ts
Comment (ludtkemorgan) [inline backend/core/src/auth/controllers/auth.controller.ts:46]: By deleting this token endpoint won't we run into the case where during the deployment all calls to the old `token` endpoint will fail. Or in other words when we deploy the backend the /auth/token endpoint will disappear so any calls to this endpoint will fail until the UI is deployed

**Review by ludtkemorgan — COMMENTED** [review]
Also, I found a doc file called `Authentication.md` that explains the authentication process. It hasn't been updated in a while, but for this change we might want to update the documentation

File: backend/core/src/auth/controllers/auth.controller.ts
Comment (YazeedLoonat) [reply backend/core/src/auth/controllers/auth.controller.ts:46]: I think that when we deploy this without a front end release things would not work either, since the front end would not have a cookie set to allow it to get through our auth checks

so I think either way we would need to have the front end and backend deployed at the same time for both of these

**Comment by YazeedLoonat** [comment]
This should be ready, the tests that are failing I'm pretty sure are more due to flakiness (looking for something that hasn't rendered quite yet)

**Review by ludtkemorgan — APPROVED** [review]

**Comment by emilyjablonski** [comment]
Pinged you offline, but having some issues signing in

**Comment by emilyjablonski** [comment]
@YazeedLoonat Looks like it's failing to compile, can you take a look so we can review?

**Comment by YazeedLoonat** [comment]
Hey @emilyjablonski can you give this another peep please I think its ready

**Comment by emilyjablonski** [comment]
@YazeedLoonat 👁️ Are the remaining issues flagged in the PR description still problems?

**Comment by YazeedLoonat** [comment]
Hey @ludtkemorgan do you have the new `CORS_ORIGINS` .env file in backend/core?

**Comment by ludtkemorgan** [comment]
> Hey @ludtkemorgan do you have the new `CORS_ORIGINS` .env file in backend/core?

Looks like my .env file was accidentally deleted so that fixed the issue!

**Comment by emilyjablonski** [comment]
https://user-images.githubusercontent.com/65367387/207986428-b5ec4af6-5587-483c-97d5-70dc64752e86.mov

Heads up, I can't log out of partners

**Comment by YazeedLoonat** [comment]
Hey @emilyjablonski can you peep one more time (hopefully) and lemme know what you think

I'm pretty sure I got this sorted out

**Review by emilyjablonski — APPROVED** [review]
So excited about this!!!

My last question is just making sure the expiry is looking good on your end?

I also would like one additional review since there have been a lot of changes since the last approval.

**Comment by ColinBuyck** [comment]
Still reviewing different cases but one that just came up is that I can log-in on the public side and then it gives me access to the partner's portal automatically. I'm assuming this is related to the admin@example account having access to both sides but unsure if it's desired behavior or a concern @YazeedLoonat 

https://user-images.githubusercontent.com/53269332/208210759-bf6e05d0-31ed-4df0-abee-498e36713ee9.mov

File: backend/core/src/auth/auth.module.ts
Comment (YazeedLoonat) [inline backend/core/src/auth/auth.module.ts:36]: flag for later

File: backend/core/.env.template
Comment (YazeedLoonat) [inline backend/core/.env.template:21]: flag for later

File: backend/core/src/auth/auth.module.ts
Comment (YazeedLoonat) [reply backend/core/src/auth/auth.module.ts:36]: https://github.com/auth0/node-jsonwebtoken/issues/609#issuecomment-504041869

**Comment by YazeedLoonat** [comment]
Hey @ColinBuyck I got some updates out in the most recent commit:
-- confirming account on public side 

-- on old access token being used both users should get logged out

are the things that got resolved 

lemme know what you think

**Review by ColinBuyck — COMMENTED** [review]
Talked offline but documenting here that I was unable to sign-in on either the public nor the partner's side on Safari with the most recent changes but did not have issues with Chrome

**Comment by YazeedLoonat** [comment]
Hey @ColinBuyck newest update should resolve this across safari, chrome and firefox lemme know what you think

**Comment by YazeedLoonat** [comment]
to add some notes the fix here was to turn off the secure flag when we are doing local development (NODE_ENV is development) 

this is because the actual request locally is done over http instead of https. Safari flags that since an http cookie can't actually be secure

this is kinda touched on: https://stackoverflow.com/a/71883080

once this gets into dev/a hosted env we can verify that the flag works properly and is marked as secure in those environments

File: backend/core/src/auth/constants.ts
Comment (ludtkemorgan) [inline backend/core/src/auth/constants.ts:11]: A comment here might be useful for explanation on why this is needed

File: backend/core/src/auth/constants.ts
Comment (ludtkemorgan) [inline backend/core/src/auth/constants.ts:8]: Maybe not for this story, but this could be a good candidate to be an environment variable in case future sites want a longer/shorter login period

**Review by ludtkemorgan — APPROVED** [review]
The most recent changes look good. I was able to login on safari! Thanks for sticking with this story.

With these changes I was getting an error when logging out. I think it was a 401 response and the UI handled it ok, but in dev mode an error popped up. However, after rebuilding and restarting I am no longer seeing that. Before merging can you just re-check the logout flow? Otherwise looks great!

**Review by ColinBuyck — APPROVED** [review]
This all looks good to me! Great work 🏆 🍪 🏆

File: backend/core/src/auth/constants.ts
Comment (YazeedLoonat) [reply backend/core/src/auth/constants.ts:8]: I agree!

**Review by YazeedLoonat — APPROVED** [review]
approving because at 1 point I requested changes

**Pull request merged**
